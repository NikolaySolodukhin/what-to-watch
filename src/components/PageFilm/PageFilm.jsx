import React, {Fragment, PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import {getFilmById, getLikeThisFilmsById, getReviews} from './../../reducer/films/selectors';
import Header from './../Header/Header.jsx';
import Footer from './../Footer/Footer.jsx';
import MovieCardInfo from './../MovieCardInfo/MovieCardInfo.jsx';
import {MovieCardFull} from './../MovieCardFull/MovieCardFull.jsx';
import CatalogMovieList from './../CatalogMovieList/CatalogMovieList.jsx';
import {Operation as OperationFilms} from "../../reducer/films/films";

class PageFilm extends PureComponent {
  componentDidMount() {
    this.props.getReviews(this.props.film.id);
  }

  render() {
    const {film, likeThisFilms, reviews} = this.props;
    return <Fragment>
      <div className="movie-card movie-card--full">
        <div className="movie-card__hero">
          <div className="movie-card__bg">
            <img src={film && film.backgroundImage} alt={film && film.name}/>
          </div>
          <h1 className="visually-hidden">WTW</h1>
          <Header withAuth />
          <MovieCardInfo film={film} />
        </div>
        <MovieCardFull film={film} reviews={reviews}/>

      </div>
      <div className="page-content">
        {likeThisFilms.length > 0 && (
          <section className="catalog catalog--like-this">
            <h2 className="catalog__title">More like this</h2>
            <CatalogMovieList filmsList={likeThisFilms}/>
          </section>
        )}
        <Footer/>
      </div>
    </Fragment>;
  }
}

PageFilm.propTypes = {
  film: PropTypes.shape({
    id: PropTypes.number.isRequired,
    backgroundColor: PropTypes.string.isRequired,
    backgroundImage: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    director: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
    isFavorite: PropTypes.bool.isRequired,
    name: PropTypes.string.isRequired,
    posterImage: PropTypes.string.isRequired,
    previewImage: PropTypes.string.isRequired,
    previewVideoLink: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
    released: PropTypes.number.isRequired,
    runTime: PropTypes.number.isRequired,
    scoresCount: PropTypes.number.isRequired,
    starring: PropTypes.array.isRequired,
    videoLink: PropTypes.string.isRequired
  }),
  likeThisFilms: PropTypes.array,
  getReviews: PropTypes.func.isRequired,
  reviews: PropTypes.arrayOf(
      PropTypes.shape({
        comment: PropTypes.string.isRequired,
        date: PropTypes.string.isRequired,
        id: PropTypes.number.isRequired,
        rating: PropTypes.number.isRequired,
        user: PropTypes.shape({
          id: PropTypes.number.isRequired,
          name: PropTypes.string.isRequired
        }).isRequired,
      }).isRequired
  ).isRequired,
};

const mapStateToProps = (state, ownProps) => ({
  ...ownProps,
  film: getFilmById(state, ownProps.match.params.id),
  likeThisFilms: getLikeThisFilmsById(state, ownProps.match.params.id),
  reviews: getReviews(state)
});

const mapDispatchToProps = (dispatch) => ({
  getReviews: (filmId) => dispatch(OperationFilms.loadReviews(filmId)),
});

export {PageFilm};

export default connect(mapStateToProps, mapDispatchToProps)(PageFilm);
