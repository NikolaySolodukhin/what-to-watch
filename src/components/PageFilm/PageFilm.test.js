import React from "react";
import Enzyme, {shallow} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import {shallowToJson} from "enzyme-to-json";

import {PageFilm} from "./PageFilm";

Enzyme.configure({adapter: new Adapter()});

describe(`<PageFilm/>`, () => {
  it(`PageFilm renders correctly`, () => {
    const mockProps = {
      film: {
        id: 1,
        genreId: 1,
        backgroundColor: `white`,
        backgroundImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
        description: `Film`,
        director: `Raccoon`,
        genre: `comedian`,
        isFavorite: false,
        name: `Fantastic Beasts`,
        posterImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
        previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
        previewVideoLink: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
        rating: 4,
        released: 2018,
        runTime: 1990,
        scoresCount: 10,
        starring: [],
        videoLink: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
      },
      reviews: [
        {
          comment: `Unfortunately we don't have a reliable way to tell the true ratings of a movie.`,
          date: `2019-06-18T20:42:47.046Z`,
          id: 1,
          rating: 2,
          user: {
            id: 10,
            name: `Max`
          }
        }
      ],
      likeThisFilms: [{
        id: 1,
        genreId: 1,
        backgroundColor: `white`,
        backgroundImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
        description: `Film`,
        director: `Raccoon`,
        genre: `comedian`,
        isFavorite: false,
        name: `Fantastic Beasts`,
        posterImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
        previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
        previewVideoLink: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
        rating: 4,
        released: 2018,
        runTime: 1990,
        scoresCount: 10,
        starring: [],
        videoLink: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
      }]
    };

    const getReviewsFunciton = jest.fn();
    const tree = shallow(
        <PageFilm
          film={mockProps.film}
          reviews={mockProps.reviews}
          likeThisFilms={mockProps.likeThisFilms}
          getReviews={getReviewsFunciton}
        />
    );
    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});
