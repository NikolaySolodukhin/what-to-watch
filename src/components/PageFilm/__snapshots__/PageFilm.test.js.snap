// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageFilm/> PageFilm renders correctly 1`] = `
<Fragment>
  <div
    className="movie-card movie-card--full"
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt="Fantastic Beasts"
          src="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
        />
      </div>
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <ConnectFunction
        withAuth={true}
      />
      <ConnectFunction
        film={
          Object {
            "backgroundColor": "white",
            "backgroundImage": "img/fantastic-beasts-the-crimes-of-grindelwald.jpg",
            "description": "Film",
            "director": "Raccoon",
            "genre": "comedian",
            "genreId": 1,
            "id": 1,
            "isFavorite": false,
            "name": "Fantastic Beasts",
            "posterImage": "img/fantastic-beasts-the-crimes-of-grindelwald.jpg",
            "previewImage": "img/fantastic-beasts-the-crimes-of-grindelwald.jpg",
            "previewVideoLink": "img/fantastic-beasts-the-crimes-of-grindelwald.jpg",
            "rating": 4,
            "released": 2018,
            "runTime": 1990,
            "scoresCount": 10,
            "starring": Array [],
            "videoLink": "https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4",
          }
        }
      />
    </div>
    <MovieCardFull
      film={
        Object {
          "backgroundColor": "white",
          "backgroundImage": "img/fantastic-beasts-the-crimes-of-grindelwald.jpg",
          "description": "Film",
          "director": "Raccoon",
          "genre": "comedian",
          "genreId": 1,
          "id": 1,
          "isFavorite": false,
          "name": "Fantastic Beasts",
          "posterImage": "img/fantastic-beasts-the-crimes-of-grindelwald.jpg",
          "previewImage": "img/fantastic-beasts-the-crimes-of-grindelwald.jpg",
          "previewVideoLink": "img/fantastic-beasts-the-crimes-of-grindelwald.jpg",
          "rating": 4,
          "released": 2018,
          "runTime": 1990,
          "scoresCount": 10,
          "starring": Array [],
          "videoLink": "https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4",
        }
      }
      reviews={
        Array [
          Object {
            "comment": "Unfortunately we don't have a reliable way to tell the true ratings of a movie.",
            "date": "2019-06-18T20:42:47.046Z",
            "id": 1,
            "rating": 2,
            "user": Object {
              "id": 10,
              "name": "Max",
            },
          },
        ]
      }
    />
  </div>
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <CatalogMoviesList
        filmsList={
          Array [
            Object {
              "backgroundColor": "white",
              "backgroundImage": "img/fantastic-beasts-the-crimes-of-grindelwald.jpg",
              "description": "Film",
              "director": "Raccoon",
              "genre": "comedian",
              "genreId": 1,
              "id": 1,
              "isFavorite": false,
              "name": "Fantastic Beasts",
              "posterImage": "img/fantastic-beasts-the-crimes-of-grindelwald.jpg",
              "previewImage": "img/fantastic-beasts-the-crimes-of-grindelwald.jpg",
              "previewVideoLink": "img/fantastic-beasts-the-crimes-of-grindelwald.jpg",
              "rating": 4,
              "released": 2018,
              "runTime": 1990,
              "scoresCount": 10,
              "starring": Array [],
              "videoLink": "https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4",
            },
          ]
        }
      />
    </section>
    <Footer />
  </div>
</Fragment>
`;
